sdk/python/kfp/components/graph_component.py

```diff
@@ -85,3 +89,165 @@ def pipeline_spec(self) -> pipeline_spec_pb2.PipelineSpec:
 
     def execute(self, **kwargs):
         raise RuntimeError('Graph component has no local execution mode.')
+
+    def _dedupe_pipeline_spec(
+        self, pipeline_spec: pipeline_spec_pb2.PipelineSpec
+    ) -> pipeline_spec_pb2.PipelineSpec:
+        """removes duplicated component spec and executor specs caused by tasks
+        calling on the same components.
+
+        component specs are only deduped when the executor specs are the
+        same, this means factors like two tasks called on the same
+        component with differing resource specs would not be deduped,
+        deduping would still occur if the caller tasks differ by their
+        input specifications since this not affect the executor spec of
+        the components
+        """
+        from dataclasses import dataclass
+
+        @dataclass
+        class clones_data_structure:
+
+            def __init__(self, component_name, component_spec, executor_name,
+                         executor_spec) -> None:
+                self.component_name = component_name
+                self.component_spec = component_spec
+                self.executor_name = executor_name
+                self.executor_spec = executor_spec
+
+            def get_data(self) -> list:
+                return [
+                    self.executor_name, self.component_spec, self.executor_spec
+                ]
+
+        components_with_clones = {}
+
+        def _collect_and_process_duplicates():
```

Comment:
Can we refactor these functions to pass in and return variables, instead of modifying non-local objects?